### ISSUE TITLE
How to make table

### ISSUE THREAD

**Issue by worifon** [gfngfn/Macrodown#6]
How to write '.mcrdh' code to make table with following data
[
  ['foo','bar','baz'],
  ['fuga','hoge','<'],
  ['^', 'fizz', 'buzz']
]

to HTML as
<table>
<tbody>
<tr><td>foo</td><td>bar</td><td>baz</td></tr>
<tr><td rowspan="2">fuga</td><td colspan="2">hoge</td></tr>
<tr><td>fizz</td><td>buzz</td></tr>
</tbody>
</table>
?

Probably, is '.mcrd' like followings?
\tbody{|
| \tr{|foo|bar|baz|}
| \tr{|fuga|hoge|<|}
| \tr{|^|fizz|buzz|}
|}

Please advise. Thank you.

**Comment by gfngfn** [comment]
Hello `worifon`, I'm sorry to answer late.

If you plan to express tables in the form like

>    [
>    ['foo','bar','baz'],
>    ['fuga','hoge','<'],
>    ['^', 'fizz', 'buzz']
>    ]

(i.e. `(string list) list`-typed data), you can define table-handling macro `\table` by the following code (in a `.mcrdh` file):

    let string-of-row row =
      match row with
      | []           -> {}
      | head :: tail -> let tail-str = string-of-row tail in
                          {<td>@head;</td>@tail-str;}

    let table-sub rowlst =
      match rowlst with
      | []          -> {}
      | row :: rest -> let row-str = string-of-row row in
                       let rest-str = table-sub rest in
                         { <tr>\deeper{@row-str;}</tr>
                           @rest-str; }

    let \table rowlst =
      let content = table-sub rowlst in {<table>\deeper{@content;}</table>}

And you cau use this macro in `.mcrd` files
like the following code (please be careful of delimiters in lists; not `,` but `;`):

    \table(
      [ [{foo}; {bar}; {baz}];
        [{fuga}; {hoge}; {<}];
        [{^}; {fizz}; {buzz}] ]
    );

or equivalently:

    \table(
      [ {| foo | bar | baz |};
        {| fuga | hoge | < |};
        {| ^ | fizz | buzz |} ]
    );

These may seem not so convienient nor elegant. But these are ones of the simplest ways,
since unfortunately current Macrodown does not have any lightweight syntax sugar for data of `(string list) list` type.

On the other hand, should you want to use tables in the form like

>    \tbody{|
>    | \tr{|foo|bar|baz|}
>    | \tr{|fuga|hoge|<|}
>    | \tr{|^|fizz|buzz|}
>    |}

(i.e. `string list`-typed data), you can define `\tbody` and `\tr` by the following code (in a `.mcrdh` file):

    let tbody rowlst =
      match rowlst with
      | []          -> {}
      | row :: tail -> let tail-str = tbody tail in { @row;
                                                      @tail-str; }

    let \tbody content = let res = tbody content in {<table>\deeper{@res;}</table>}

    let tr tdlst =
      match tdlst with
      | []         -> {}
      | td :: tail -> let tail-str = tr tail in { <td>@td;</td>
                                                  @tail-str;}

    let \tr content = let res = tr content in {<tr>\deeper{@res;}</tr>}

And you can use this in `.mcrd` files like

    \tbody{
    | \tr{|foo|bar|baz|}
    | \tr{|fuga|hoge|<|}
    | \tr{|^|fizz|buzz|}
    |}

This implementation is, however, somewhat “awkward”, for you can write in contents of the other form than `\tr{|…|…|…|}`.

In conclusion, there are some ways for writing tables in Macrodown as I wrote above, but they are not natural, and current Macrodown does not have lightweight syntax sugar for writing tables. This problem is left for our future work.

I'm delighted to have been asked the important question. Thank you.

**Comment by worifon** [comment]
Thank you very much for great explanation.
I expect you great future work.
